From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Boris Krasnovskiy <boris.krasnovskiy@ezurio.com>
Date: Wed, 15 Jul 2026 00:00:00 +0000
Subject: [PATCH] media: imx8-media-dev: defer when sensor is not bound and fix
 link setup

When sensor drivers have not finished probing/registering when
imx8-media-dev loads, imx8-media-dev would unregister the ISI channel
and destroy the media graph. Defer probe until the sensor device has
bound.

Additionally:
- Get sensor source pad index by parsing the DT connection graph.
- Remove link_setup call on sensor subdevices that do not implement it,
  and tolerate -ENOIOCTLCMD on media link setup calls.

Signed-off-by: Boris Krasnovskiy <boris.krasnovskiy@ezurio.com>
Upstream-Status: Inappropriate NXP proprietary driver
---
 drivers/staging/media/imx/imx8-media-dev.c | 56 ++++++++++++++++++----
 1 file changed, 46 insertions(+), 10 deletions(-)

diff --git a/drivers/staging/media/imx/imx8-media-dev.c b/drivers/staging/media/imx/imx8-media-dev.c
index 534fa5a4cd14..d7e6d0971d8e 100644
--- a/drivers/staging/media/imx/imx8-media-dev.c
+++ b/drivers/staging/media/imx/imx8-media-dev.c
@@ -128,6 +128,10 @@ struct mxc_md {
 	struct v4l2_async_notifier subdev_notifier;
 };
 
+static int debug;
+module_param(debug, int, 0644);
+MODULE_PARM_DESC(debug, "Debug level (0-2)");
+
 static inline struct mxc_md *notifier_to_mxc_md(struct v4l2_async_notifier *n)
 {
 	return container_of(n, struct mxc_md, subdev_notifier);
@@ -280,7 +284,7 @@ static int mxc_md_clean_unlink_channels(struct mxc_md *mxc_md)
 	return 0;
 }
 
-static void mxc_md_unregister_all(struct mxc_md *mxc_md)
+__maybe_unused static void mxc_md_unregister_all(struct mxc_md *mxc_md)
 {
 	struct mxc_isi_info *mxc_isi;
 	int i;
@@ -299,6 +303,8 @@ static void mxc_md_unregister_all(struct mxc_md *mxc_md)
 
 static int mxc_md_create_links(struct mxc_md *mxc_md)
 {
+	struct device_node *csi_ep, *remote_ep;
+	struct of_endpoint endpoint;
 	struct media_entity *source, *sink;
 	struct mxc_isi_info *mxc_isi;
 	struct mxc_sensor_info *sensor;
@@ -351,6 +357,8 @@ static int mxc_md_create_links(struct mxc_md *mxc_md)
 		/* Notify capture subdev entity ,ISI cap link setup */
 		ret = media_entity_call(source, link_setup, &source->pads[source_pad],
 					&sink->pads[sink_pad], flags);
+		if (ret == -ENOIOCTLCMD)
+			ret = 0;
 		if (ret) {
 			v4l2_err(&mxc_md->v4l2_dev,
 				 "failed call link_setup [%s] %c> [%s]\n",
@@ -458,6 +466,8 @@ static int mxc_md_create_links(struct mxc_md *mxc_md)
 		ret = media_entity_call(sink, link_setup,
 					&sink->pads[sink_pad],
 					&source->pads[source_pad], 0);
+		if (ret == -ENOIOCTLCMD)
+			ret = 0;
 		if (ret)
 			break;
 
@@ -465,6 +475,8 @@ static int mxc_md_create_links(struct mxc_md *mxc_md)
 		ret = media_entity_call(source, link_setup,
 					&source->pads[source_pad],
 					&sink->pads[sink_pad], 0);
+		if (ret == -ENOIOCTLCMD)
+			ret = 0;
 		if (ret)
 			break;
 
@@ -501,16 +513,11 @@ static int mxc_md_create_links(struct mxc_md *mxc_md)
 			ret = media_entity_call(sink, link_setup,
 						&sink->pads[sink_pad],
 						&source->pads[source_pad], 0);
+			if (ret == -ENOIOCTLCMD)
+				ret = 0;
 			if (ret)
 				return ret;
 
-			/* Notify MIPI sensor subdev entity */
-			ret = media_entity_call(source, link_setup,
-						&source->pads[source_pad],
-						&sink->pads[sink_pad],
-						0);
-			if (ret)
-				return ret;
 			v4l2_info(&mxc_md->v4l2_dev,
 				  "created link [%s] => [%s]\n",
 				  source->name, sink->name);
@@ -519,7 +526,31 @@ static int mxc_md_create_links(struct mxc_md *mxc_md)
 
 			source = &sensor->sd->entity;
 			sink = find_entity_by_name(mxc_md, mipi_csi2->sd_name);
-			source_pad = 0;
+			csi_ep = of_graph_get_next_endpoint(mipi_csi2->node, NULL);
+			if (!csi_ep) {
+				v4l2_err(&mxc_md->v4l2_dev,
+					 "Failed to get CSI endpoint\n");
+				return -ENODEV;
+			}
+
+			remote_ep = of_graph_get_remote_endpoint(csi_ep);
+			of_node_put(csi_ep);
+			if (!remote_ep) {
+				v4l2_err(&mxc_md->v4l2_dev,
+					 "Failed to get CSI remote endpoint\n");
+				return -ENODEV;
+			}
+
+			memset(&endpoint, 0x0, sizeof(struct of_endpoint));
+			ret = of_graph_parse_endpoint(remote_ep, &endpoint);
+			of_node_put(remote_ep);
+			if (ret < 0) {
+				v4l2_err(&mxc_md->v4l2_dev,
+					 "Failed to parse remote endpoint\n");
+				return ret;
+			}
+
+			source_pad = endpoint.port;
 			sink_pad = source_pad;
 
 			mipi_vc = (mipi_csi2->vchannel) ? 4 : 1;
@@ -538,14 +569,8 @@ static int mxc_md_create_links(struct mxc_md *mxc_md)
 							&sink->pads[sink_pad + j],
 							&source->pads[source_pad + j],
 							0);
-				if (ret)
-					return ret;
-
-				/* Notify MIPI sensor subdev entity */
-				ret = media_entity_call(source, link_setup,
-							&source->pads[source_pad + j],
-							&sink->pads[sink_pad + j],
-							0);
+				if (ret == -ENOIOCTLCMD)
+					ret = 0;
 				if (ret)
 					return ret;
 			}
@@ -1026,10 +1051,10 @@ static int register_sensor_entities(struct mxc_md *mxc_md)
 		 * Need to wait sensor driver probed for the first time
 		 */
 		client = of_find_i2c_device_by_node(rem);
-		if (!client) {
-			v4l2_info(&mxc_md->v4l2_dev,
-				  "Can't find i2c client device for %s\n",
-				  of_node_full_name(rem));
+		if (!client || !client->dev.driver || !dev_get_drvdata(&client->dev)) {
+			v4l2_dbg(1, debug, &mxc_md->v4l2_dev,
+				 "Can't find i2c client device for %s\n",
+				 of_node_full_name(rem));
 			return -EPROBE_DEFER;
 		}
 
@@ -1116,10 +1141,9 @@ static int mxc_md_probe(struct platform_device *pdev)
 
 				mxc_md_clean_unlink_channels(mxc_md);
 			} else {
-				/* no sensors connected */
-				mxc_md_unregister_all(mxc_md);
-				v4l2_async_nf_unregister(&mxc_md->subdev_notifier);
-				v4l2_async_nf_cleanup(&mxc_md->subdev_notifier);
+				/* sensor not bound yet, defer probe */
+				ret = -EPROBE_DEFER;
+				goto err_register_nf;
 			}
 		}
 	}
-- 
2.43.0
